cpp/src/arrow/array/validate.cc

```diff
@@ -454,9 +437,36 @@ struct ValidateArrayImpl {
     return data.buffers[index] != nullptr && data.buffers[index]->address() != 0;
   }
 
-  Status RecurseInto(const ArrayData& related_data) {
-    ValidateArrayImpl impl{related_data, full_validation};
-    return impl.Validate();
+  template <typename... FieldDescription>
+  Status RecurseIntoField(int field_index, const FieldDescription&... description) {
+    const auto& related_data = *data.child_data[field_index];
+
+    return RecurseInto(related_data,
+                       data.type->storage_type()->field(field_index)->nullable(),
+                       description...);
```

Comment:
This may not be performance-critical, but is there any reason not to use universal refs and std::forward here?

Reply:
Nothing takes ownership of any of the arguments, so taking them by const ref doesn't cause any copies. The reason not to use std::forward was just less boilerplate, but I can add it if you'd prefer consistency

Reply:
No, I was just wondering. It's fine with me like this.